src/cargo/ops/registry/publish.rs

```diff
@@ -136,6 +136,35 @@ pub fn publish(ws: &Workspace<'_>, opts: &PublishOpts<'_>) -> CargoResult<()> {
     )?;
     verify_dependencies(pkg, &registry, reg_ids.original)?;
 
+    // Bail before packaging and uploading if same version already exists in the registry
+
+    let mut source = SourceConfigMap::empty(opts.gctx)?.load(reg_ids.original, &HashSet::new())?;
+
+    let query = Dependency::parse(pkg.name(), Some(&ver), reg_ids.original)?;
+
+    let _lock = opts
+        .gctx
+        .acquire_package_cache_lock(CacheLockMode::DownloadExclusive)?;
+
+    let duplicate_query = loop {
+        match source.query_vec(&query, QueryKind::Exact) {
+            std::task::Poll::Ready(res) => {
+                break res?;
+            }
+            std::task::Poll::Pending => source.block_until_ready()?,
+        }
+    };
+
+    drop(_lock);
+
+    if !duplicate_query.is_empty() {
+        bail!(
+            "crate {} already has version {}. Aborting publish.",
```

Comment:
nit:
`"crate {name}@{version} is already present in {registry}"`
*(have not looked into what "registery" variable would be appropriate to lit here)*

- Specifying the registry might be helpful in case someone is publishing to multiple as that gives them a chance to think "Oh whoops, I meant this other one"
- We likely don't need an "Aborting publish" as thats what an error is
- The current phrasing is a little awkward (saying a crate has a version reads a little funny to me)

Reply:
registry variable here can be `source.describe()` which would be "crates.io index" in this case.